app/services/get_products_service.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+class GetProductsService < BaseService
+  def initialize(category_id)
+    super()
+    @category_id = category_id
+  end
+
+  def call
+    validate
+    return self if error?
+
+    query_statement = build_query_statement
+    @result = fetch_results(query_statement)
+    result
+  end
+
+  private
+
+  attr_reader :category_id
+
+  def build_query_statement
+    if Category.where(id: @category_id, parent_id: nil).present?
+      Category.with(categories: Category.left_outer_joins(:parent))
+    else
+      Category.with(categories: Category.left_outer_joins(:parent)
+                                        .where(parent_id: @category_id)
+                                        .or(Category.where(id: @category_id)))
+    end
+  end
+
+  def fetch_results(query_statement)
```

Comment:
why we need this method ? this method continue build a query statement and does not fetch anything, u can test it in rails c

```ruby
def build_category_cte
    if Category.where(id: @category_id, parent_id: nil).present?
      Category.with(categories: Category.left_outer_joins(:parent))
    else
      Category.with(categories: Category.left_outer_joins(:parent)
                                        .where(parent_id: @category_id)
                                        .or(Category.where(id: @category_id)))
    end
  end

  def build_query
    category_cte = build_category_cte
    category_cte
      .joins(:products)
      .left_outer_joins(products: :tag)
      .select('categories.id AS category_id, categories.name AS category_name,
             products.id AS product_id, products.name AS product_name,
             products.price AS product_price, products.thumbnail AS product_thumbnail,
             tags.name AS tag_name, tags.color AS tag_color')
  end
  ```

Reply:
It is changed after the issue was pointed out